Instant Search: Add support for hiding search sort selector

jsnmoon opened the pull request:
Split out from #15727.

Fixes #15735 

#### Changes proposed in this Pull Request:
* Adds a site-wide option to hide the sort selector in the Jetpack Search overlay.
* Removes the `Show search box` option from the Jetpack Search widget when instant search is enabled.
* Removes the `Show sort selection dropdown` option from the Jetpack Search widget when instant search is enabled. 
* Rewords various Jetpack Search customizer section descriptions for brevity and consistency.
* Adds a custom `Label_Control` class to be used in the Jetpack Search customizer section.
* Refactors the bridging logic between the Customizer and the Instant Search app to avoid duplicating code.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Yes, it adds the ability to hide the sort selector in the Jetpack Search overlay. The search overlay is only available to customers with the Search product plan.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Apply this change to a site with Jetpack Instant Search enabled.

2. Add a Jetpack Search widget to your site's footer or sidebar in the customizer.

3. Ensure that neither `Show search box` nor `Show sort selection dropdown` checkboxes are available. 

> <img width="292" alt="Screen Shot 2020-05-28 at 5 33 35 PM" src="https://user-images.githubusercontent.com/4044428/83204576-6507a980-a109-11ea-894c-9875d9db2a00.png">

4. Ensure that the widget title, post types checkboxes, default sort order, and filter controls work as expected. These options should not be affected by this change.

5. Navigate to the Jetpack Search section of the customizer. 

6. Ensure there's a group of checkboxes titled `Additional Jetpack Search Settings`. 

> <img width="292" alt="Screen Shot 2020-05-28 at 5 35 55 PM" src="https://user-images.githubusercontent.com/4044428/83204699-b44dda00-a109-11ea-9b3e-d32cca6ac4a5.png">

7. Try toggling the newly added `Show Sort Selector` checkbox. Verify that unchecking the checkbox hides the sort selector in the search overlay without refreshing the page.

> <img width="686" alt="Screen Shot 2020-05-11 at 3 13 39 PM" src="https://user-images.githubusercontent.com/4044428/81612412-036dde00-939a-11ea-9398-9915a1721762.png">
> <img width="686" alt="Screen Shot 2020-05-11 at 3 14 36 PM" src="https://user-images.githubusercontent.com/4044428/81612478-239d9d00-939a-11ea-86eb-a27c069dd344.png">

8. Try disabling instant search via `/wp-admin/admin.php?page=jetpack#/performance` and repeat step 3. Both `Show search box` and `Show sort selection dropdown` checkboxes should be available again.

#### Proposed changelog entry for your changes:
* Adds a Customizer control for hiding the sort selector in the Jetpack Search overlay.

Commit:
Add support for hiding search sort selector

Commit:
Restore sort control checkbox to widget form

Commit:
Create separate fn's for handling Instant Search

Remove search_box_enabled option

Commit:
Undo custom widget update logic for instant search

Commit:
Undo changes related to sort property propagation

Commit:
Fix sprintf invocation

bluefuton commented on modules/search/class-jetpack-instant-search.php:
```diff
@@ -8,6 +8,7 @@
 
 use Automattic\Jetpack\Connection\Client;
 use Automattic\Jetpack\Constants;
+require_once dirname( __FILE__ ) . '/class.jetpack-search-helpers.php';
```
You can use `__DIR__` instead of `dirname( __FILE__)` :)

bluefuton commented on modules/search/customize-controls/class-label-control.php:
```diff
@@ -0,0 +1,49 @@
+	 * Enqueue styles related to this control.
+	 */
+	public function enqueue() {
+		require_once dirname( dirname( __FILE__ ) ) . '/class.jetpack-search-helpers.php';
```
You can use `dirname( __FILE__, 2 )` to avoid repeating dirname 👍

jsnmoon replied:
According to PHPCS, `dirname()` doesn't have a "levels" parameter in PHP version 5.6 or earlier. Since Jetpack supports PHP 5.6, I'll keep this as-is.

bluefuton commented:
I think the checkbox labels should be in sentence case, to match the rest of the Customiser. For example:

<img width="228" alt="Screen Shot 2020-06-17 at 14 22 38" src="https://user-images.githubusercontent.com/17325/84848045-8293b780-b0a6-11ea-9978-dc71800bda40.png">

bluefuton commented:
Great testing instructions (as always).

On my test site, turning off the 'Show sort selector' checkbox made the sort selector disappear in the Customiser preview 👍

However, the setting did not seem to persist, and the sort selector continued appearing in the search on the site front end. When I revisit the Customiser, the 'Show sort selector' box is checked again.

Commit:
Merge master into add/search-toggle-sort-control

Commit:
Address review feedback

Commit:
Sanitize boolean-like values

jsnmoon commented:
> I think the checkbox labels should be in sentence case, to match the rest of the Customiser. @bluefuton 

✅ 

> However, the setting did not seem to persist, and the sort selector continued appearing in the search on the site front end. When I revisit the Customiser, the 'Show sort selector' box is checked again. @bluefuton 

I found an issue where checkbox customizer values weren't being properly sanitized before being written to the DB. This should be fixed.

bluefuton reviewed (dismissed):
The setting now persists correctly 👍 Looks good!

Also seems to have fixed the 'powered by Jetpack' toggle (https://github.com/Automattic/jetpack/issues/15735).

:shipit:

jeherve reviewed (changes requested):
I think there was a bit of a mishap during the rebase, the PR now includes unrelated changes?

jsnmoon commented:
> I think there was a bit of a mishap during the rebase, the PR now includes unrelated changes?

🤦 You're right, fixing now.

Commit:
Merge branch master into add/search-toggle-sort-control

Commit:
[not verified] Revert unintended changes from merge

jsnmoon commented:
@jeherve: Should be fixed, thanks!

jeherve commented on modules/search/class-jetpack-instant-search.php:
```diff
@@ -8,6 +8,7 @@
 
 use Automattic\Jetpack\Connection\Client;
 use Automattic\Jetpack\Constants;
+require_once __DIR__ . '/class.jetpack-search-helpers.php';
```
Do we need this here, even though the file is already required by the base class?

jsnmoon replied:
Ah, good point; this require is unnecessary.

jeherve commented on modules/search/class-jetpack-search-customize.php:
```diff
@@ -10,7 +10,9 @@
 	exit;
 }
 
+require_once dirname( __FILE__ ) . '/class.jetpack-search-helpers.php';
```
Do we need the require here?

jsnmoon replied:
No, although it was necessary at one point during development. I'll remove this.

jeherve commented on modules/search/customize-controls/class-label-control.css:
```diff
@@ -0,0 +1,18 @@
+#sub-accordion-section-jetpack_search .customize-control {
+  margin-bottom: 14px;
```
You'll probably want to move from spaces to tabs, to avoid someone accidentally changing all those at once on save when they next edit this file.

jeherve commented on modules/search/customize-controls/class-label-control.php:
```diff
@@ -0,0 +1,49 @@
+		<?php endif;
+	}
+}
+?>
```
Can we do without the closing tag here?

jeherve reviewed (dismissed):
This seems problematic on sites that already have a widget set up; the sort selector is then hidden by default on the frontend once you update to this branch. I think it should remain on by default in those cases.

Other than that, I only have minor remarks below.

Commit:
Remove unnecessary requires

Commit:
Use tabs for indentation

Commit:
Remove closing tag, fix indentation

Commit:
Fix checkbox customizer config once and for all

jsnmoon commented:
> This seems problematic on sites that already have a widget set up; the sort selector is then hidden by default on the frontend once you update to this branch. I think it should remain on by default in those cases. @jeherve 

Agreed, it looks like my fix overlooked this scenario. Here are what `get_option` for these checkbox customizer values can return:

> Truthy values:
> - `'1'` (after being set via the Customizer)
> - `true` (before customizer write, in customizer iframe, when default set to `true`)
> - `x` (before customizer write, in customizer iframe, when default set to `x`)
> 
> Falsy values:
> - `''` (after being set via the Customizer)
> - `false` (before customizer write, outside the customizer iframe, regardless of default value)

I went ahead and changed the default setting values for these checkboxes to `'1'` and added a `=== '1'` comparison when getting these option values. This approach should fix the issue for good.

gibrown commented:
@jsnmoon will this also fix https://github.com/Automattic/jetpack/issues/16123 ? While deploying onto support sites I've had a lot of questions about the default sorting order before the user has searched.

jeherve reviewed (approved):
This works well in my tests now. 🚢

jsnmoon commented:
> will this also fix #16123? While deploying onto support sites I've had a lot of questions about the default sorting order before the user has searched. @gibrown 

This doesn't address #16123; #16017 might, though.
